File: src/main/java/net/miarma/mkernel/commands/misc/LobbyCommand.java
package net.miarma.mkernel.commands.misc;

import java.util.List;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.entity.Player;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.MKernel;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.config.providers.ConfigProvider;
import net.miarma.mkernel.config.providers.MessageProvider;
import net.miarma.mkernel.util.MessageUtil;

public class LobbyCommand {
    public static void register() {
        CommandWrapper lobbyCmd = CommandProvider.getLobbyCommand();
        new CommandAPICommand(lobbyCmd.getName())
            .withFullDescription(lobbyCmd.getDescription())
            .withShortDescription(lobbyCmd.getDescription())
            .withPermission(lobbyCmd.getPermission().base())
            .withOptionalArguments(
            		CommandProvider.Arguments.playersOptArg().withPermission(
                    lobbyCmd.getPermission().others()
                )
            )
            .executesPlayer((sender,args) -> {
                boolean lobbyExists = Bukkit.getServer().getWorlds().stream()
                        .map(World::getName)
                        .map(String::toLowerCase)
                        .anyMatch(w -> w.contains(ConfigProvider.Worlds.getLobby().name()));

                if(lobbyExists) {
                    String name = ConfigProvider.Worlds.getLobby().name();
                    double x = ConfigProvider.Worlds.getLobby().x();
                    double y = ConfigProvider.Worlds.getLobby().y();
                    double z = ConfigProvider.Worlds.getLobby().z();
                    int yaw = ConfigProvider.Worlds.getLobby().yaw();
                    int pitch = ConfigProvider.Worlds.getLobby().pitch();
                    Location lobbyCoords = new Location(Bukkit.getWorld(name), x, y, z, yaw, pitch);

                    if (args.count() == 0) {
                        sender.teleport(lobbyCoords);
                        MessageUtil.sendMessage(sender, lobbyCmd.getMessages()[0], true);
                    } else if (args.count() >= 1) {
                        Player victim = Bukkit.getServer().getPlayer(args.getRaw(0));
                        victim.teleport(lobbyCoords);
                        MessageUtil.sendMessage(sender, lobbyCmd.getMessages()[1], true,
                                                    List.of("%victim%"), List.of(victim.getName()));
                        MessageUtil.sendMessage(victim, lobbyCmd.getMessages()[2], true,
                                                    List.of("%sender%"), List.of(sender.getName()));
                    }
                } else {
                    MessageUtil.sendMessage(sender, MessageProvider.Errors.lobbyDoesNotExist(), true);
                    MKernel.LOGGER.warning(MessageUtil.formatMessageConsole(
                            MessageProvider.Errors.lobbyDoesNotExist(), true));
                }
            })
            .register();
    }
}

File: src/main/kotlin/net/miarma/mkernel/command/impl/misc/LobbyCommand.kt
package net.miarma.mkernel.command.impl.misc

import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.arguments.PlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.command.MCommand
import net.miarma.mkernel.common.config.ConfigKeys
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.MessageService
import net.miarma.mkernel.util.PlayerUtil
import java.util.logging.Logger

@Singleton
class LobbyCommand @Inject constructor(
    private val configService: ConfigService,
    private val messageService: MessageService,
    private val logger: Logger
) : MCommand {
    override fun register() {
        commandAPICommand(configService.getString(ConfigKeys.Commands.Lobby.NAME)) {
            withFullDescription(configService.getString(ConfigKeys.Commands.Lobby.DESC))
            withPermission(configService.getString(ConfigKeys.Commands.Lobby.PERM_BASE))
            withOptionalArguments(
                PlayerProfileArgument(configService.getString(ConfigKeys.Arguments.PLAYER))
                    .withPermission(configService.getString(ConfigKeys.Commands.Lobby.PERM_OTHERS))
            )
            playerExecutor { sender, args ->
                val lobby = configService.getLobbyWorld()
                val lobbyCoords = lobby.toLocation()

                if (lobbyCoords.world == null) {
                    messageService.builder(configService.getString(ConfigKeys.Messages.Teleport.Errors.LOBBY_NOT_EXIST)).withPrefix().send(sender)
                    logger.warning("The lobby defined in config does not exist!")
                    return@playerExecutor
                }

                if (args[0] == null) {
                    sender.teleportAsync(lobbyCoords).thenAccept { success ->
                        if (success) messageService.builder(configService.getString(ConfigKeys.Commands.Lobby.MSG_TELEPORTED)).withPrefix().send(sender)
                    }
                    return@playerExecutor
                }

                val target = PlayerUtil.fromArg(args[0])
                if (target == null || !target.isOnline) {
                    messageService.builder(configService.getString(ConfigKeys.Messages.General.Errors.PLAYER_NOT_FOUND)).withPrefix().send(sender)
                    return@playerExecutor
                }

                target.teleportAsync(lobbyCoords).thenAccept { success ->
                    if (success) {
                        messageService.builder(configService.getString(ConfigKeys.Commands.Lobby.MSG_YOU_OTHERS)).withPrefix().tag("target", target.name).send(sender)
                        messageService.builder(configService.getString(ConfigKeys.Commands.Lobby.MSG_OTHERS_YOU)).withPrefix().forPlayer(target).tag("sender", sender.name).send(target)
                    }
                }
            }
        }
    }
}